Public/Src/Tools/Execution.Analyzer/Analyzers.Core/XLGPlusPlus/XLGToDBAnalyzer.cs

```diff
@@ -59,103 +55,360 @@ private static void WriteXLGToDBHelp(HelpWriter writer)
+            m_actionBlock.Post(() =>
+            {
+                m_inner.ProcessEvent(data, workerId);
+            });
+        }
+
+        /// <inheritdoc/>
+        public override void Prepare()
+        {
+            m_inner.Prepare();
+        }
+
+        /// <inheritdoc/>
+        public override int Analyze()
+        {
+            m_actionBlock.Complete();
+            m_actionBlock.CompletionAsync().GetAwaiter().GetResult();
+            return m_inner.Analyze();
+        }
+
+        /// <inheritdoc/>
+        public override void Dispose()
+        {
+            m_inner.Dispose();
+        }
+    }
+
 
     /// <summary>
     /// Analyzer to dump xlg events and other data into RocksDB
     /// </summary>
-    internal sealed class XLGToDBAnalyzer : Analyzer
+    internal sealed class XLGToDBAnalyzerInner : Analyzer
     {
+        /// <summary>
+        /// Output directory path
+        /// </summary>
         public string OutputDirPath;
+
+        /// <summary>
+        /// Store WorkerID.Value to pass into protobuf object to identify this event
+        /// </summary>
+        public ThreadLocal<uint> WorkerID = new ThreadLocal<uint>();
+
         private bool m_accessorSucceeded;
-        private BXLInvocationEventList m_invocationEventList = new BXLInvocationEventList();
-        private KeyValueStoreAccessor Accessor { get; set; }
-        private uint WorkerID { get; set; }
 
+        private KeyValueStoreAccessor m_accessor;
+
+        private Stopwatch m_stopWatch;
+
+        private int m_eventCount;
+
+        public XLGToDBAnalyzerInner(AnalysisInput input) : base(input)
+        {
+            m_stopWatch = new Stopwatch();
+            m_stopWatch.Start();
+        }
+
+        internal void ProcessEvent<TEventData>(TEventData data, uint workerId) where TEventData : struct, IExecutionLogEventData<TEventData>
+        {
+            var eventCount = Interlocked.Increment(ref m_eventCount);
+
+            if (eventCount % 10000 == 0)
+            {
+                Console.WriteLine("Processed {0} events so far.", eventCount);
+                Console.WriteLine("Total time elapsed: {0} seconds", m_stopWatch.ElapsedMilliseconds / 1000.0);
+            }
 
-        public XLGToDBAnalyzer(AnalysisInput input) : base(input) { }
+            WorkerID.Value = workerId;
+            data.Metadata.LogToTarget(data, this);
+        }
 
         /// <inheritdoc/>
         public override void Prepare()
         {
-            try
+            // If a directory exists and it contains sst files (meaning it was a previous RocksDB instance), delete the directory
+            // Also this way you don't accidentally nuke a directory that you passed in, but wasn't a previous RocksDB instance.
+            if (Directory.Exists(OutputDirPath) && Directory.GetFiles(OutputDirPath, "*.sst").Length > 0)
             {
                 Directory.Delete(path: OutputDirPath, recursive: true);
             }

```

Comment:
you might have a dir with initialized but empty RocksDB - there will be files but there will not be sst files

Reply:
I'd check whether a directory exists and whether it's empty. if it does exists, and there are files in it - show an error and abort

---
In reply to: [309373958](https://github.com/microsoft/BuildXL/pull/671#discussion_r309373958) [](ancestors = 309373958)

Reply:
That's fine, RocksDb will just clean out any existing files (and append the content/metadata to its log). 

---
In reply to: [309373958](https://github.com/microsoft/BuildXL/pull/671#discussion_r309373958) [](ancestors = 309373958)